NotificationServiceExtension/NotificationService.swift

```diff
@@ -2,57 +2,178 @@
         if let contentHandler = contentHandler,
            let bestAttemptContent =  bestAttemptContent {
             contentHandler(bestAttemptContent)
         }
     }
+    
+    private func handleTalkPageTypeNotificationsIfNeeded(notifications: Set<RemoteNotificationsAPIController.NotificationsResult.Notification>, bestAttemptContent: UNMutableNotificationContent, contentHandler: ((UNNotificationContent) -> Void)) -> Bool {
+        
+        if allNotificationsAreForSameTalkPage(notifications: notifications) {
+            let messageText = String.localizedStringWithFormat(WMFLocalizedString("notifications-push-talk-messages-format", value: "{{PLURAL:%1$d|message|messages}}", comment: "Plural messages text to be inserted into push notification content - %1$d is replaced with the number of talk page messages."), notifications.count)
+            bestAttemptContent.subtitle = String.localizedStringWithFormat(WMFLocalizedString("notifications-push-talk-title-format", value: "New %1$@", comment: "Title text for a push notification that represents talk page messages - %1$@ is replaced with \"Messages\" text (can be plural or singular)."), messageText)
+            
+            if notifications.count == 1 {
+                guard let pushContentText = notifications.first?.pushContentText else {
+                    contentHandler(bestAttemptContent)
+                    return true
+                }
+                
+                bestAttemptContent.body = pushContentText
+                contentHandler(bestAttemptContent)
+                return true
+            } else {
+                
+                guard let talkPageTitle = notifications.first?.titleFull else {
+                    contentHandler(bestAttemptContent)
+                    return true
+                }
+                
+                bestAttemptContent.body = String.localizedStringWithFormat(WMFLocalizedString("notifications-push-talk-body-format", value: "%1$d new %2$@ on %3$@", comment: "Body text for a push notification that represents talk page messages - %1$d is replaced with the number of talk page messages, %2$@ is replaced with \"messages\" text (can be plural or singular), and %3$@ is replaced with the talk page title. For example, \"3 new messages on User talk: Username\""), notifications.count, messageText, talkPageTitle)
+                contentHandler(bestAttemptContent)
+                return true
+            }
+        }
+        
+        return false
+    }
+    
+    private func allNotificationsAreForSameTalkPage(notifications: Set<RemoteNotificationsAPIController.NotificationsResult.Notification>) -> Bool {
+        
+        guard notifications.count > 0 else {
+            return false
+        }
+        
+        typealias TalkPageName = String
+        typealias NotificationKey = String
+        var talkDictionary: [TalkPageName: [NotificationKey]] = [:]
+        
```

Comment:
Could this be more performant if we did something like


```
guard let firstNotificationTitle = notifications.first.titleFull else {
return false
}

for notification in notifications {
if notification.titleFull != firstNotificationTitle {
return false
}

return true
}
```

Reply:
@mcleinman yes! Good point, I was definitely overcomplicating it. I will say this method should rarely be passed in very many notifications because we should be filtering out any > 10 minutes ago before this point, but it was still pretty unreadable before. I updated this in the new helper class.